Add guide and checklist for reviewing release PRs (#1976)
MetaMask/core: docs/reviewing-release-prs.md

# Reviewing Release PRs

In order to publish a new release of the monorepo, a release PR must be created. But before a release PR is merged, it is absolutely critical — even if you are the author of the PR — to review the changes that are about to be published and to ensure that consumers know about them effectively.

This document shows you how to do that.

## 0. tl;dr — if you've done this before

- [ ] Are there any packages with version bumps that should not be published?
- [ ] Do the new versions of packages being published sufficiently communicate the impact of changes to those packages?
- [ ] Are all changelog entries placed in the correct category?
- [ ] Do all changelog entries explain the changes in a clear and straightforward manner and detail exact changes to the package's interface?
- [ ] Are there any changelog entries that do not need to be listed?
- [ ] Are there any changes made to a package since its previous release that are not reflected in its changelog?
- [ ] Are there any changelog entries that have missing references to pull requests?
- [ ] Have any changes been merged to `main` while the release PR has been open that have not been added to a changelog?

> - [ ] Are there any bumps to internal dependencies that have not been captured in changelogs?

Author: Yes, good call.

> - [ ] Are there any bumps to internal dependencies that have not been captured in changelogs?

Author: I looked at this again and I added this already under "Are there any changelog entries that have missing references to pull requests?". I can reword it and move it up, though, to make it more clear.

> - [ ] Are there any bumps to internal dependencies that have not been captured in changelogs?

Author: Right, good call.

> - [ ] Are there any bumps to internal dependencies that have not been captured in changelogs?

Author: Updated here: [1b3ce09](https://github.com/MetaMask/core/pull/1976/commits/1b3ce09c8431dcc316ed76236215745772b805b1). Hopefully that is a bit more clear but let me know if it's not!

- Changes to interfaces
- Changes to behavior
- Changes to runtime dependencies
- Changes to consumer-facing documentation

Conversely, you generally should not need to worry about:

- Changes to internal tooling
- Changes to developer-only dependencies
- Changes to contributor-only documentation

With that in mind, there are four categories of changes that determine the level of impact.

### Breaking changes

If a consumer upgrades a package to a new version without making any modifications to their code, that version is "breaking" if it causes:

- An error at runtime
- An error at compile time (e.g., a TypeScript type error)
- An error at install time (e.g., the consumer's package manager reports a Node version incompatibility)
- A surprising difference in behavior

Given this, there are many ways a change could be regarded as breaking:

- Changing a function or method to throw an error
- Adding a required argument to a function or method
- Adding a property to a TypeScript type
- Narrowing the type of a property in a TypeScript type
- Narrowing the type of an argument in a function
- Adding a parameter to a TypeScript type
- Removing a parameter from a TypeScript type

> - Upgrading a runtime dependency to a version which causes any of the above

Author: This one is very important. I could go into more detail about how to discover this.

Author: I won't have enough time this sprint to do this, but am happy to do this in a future sprint.

- Removing an export from the package
- Renaming the package
- Bumping the Node version
- Bumping the NPM version

### Additions

Minimally, this category could cover any actions that extend the surface area of a package in some way:

- Adding an optional property to an existing type
- Adding an optional argument to an existing function or method
- Adding a new export to a package

However, it is important to remember that consumers care about capabilities first and interfaces second. So, some additions could also be more simply regarded as changes. For instance:

- Adding a new dependency

### Removals

In most cases, removing something from a package will cause a breaking change, and thus should be categorized as such. There are some rare instances, however, when this is not the case:

- Removing a runtime dependency

### Fixes

There are a few different kinds of changes that could be filed under this category.

> - Fixing a security vulnerability. (NOTE: These are actually a special kind of change and should be made discreetly through a private channel so as not to raise suspicion.)

Author: That's a good point. I will remove that statement.

Therefore, if there are any changelog entries that should be moved to a different category, or if there are any changelog entries that can be reworded to help consumers, or if there are changelog entries that are missing pull request references, notify the creator of the release PR, offering a suggestion as necessary.

## 5. Add missing changelog entries

Finally, look over the changelogs one more time to ensure that they aren't missing anything. There are three ways that entries could get omitted:

- Although the `create-release-branch` tool helps to autopopulate changelog entries, it does not remove the human element entirely, and it is possible that the creator of the release PR may have inadvertently removed some entries before pushing up the branch.
- While a release PR is open, more changes may be made to packages that are already planned for release.
- Finally, if the version of a package in the monorepo is updated, the manifests of any dependents of that package will be updated so that they match the same version. This is unusual because bumps to dependencies usually happen _before_ a release PR, but is essential to document just the same.

> Therefore, review the [changes committed since the previous release](#review-changes) and look for internal dependency bumps, and if either are missing from the changelogs, notify the creator of the release PR so that they can add them, offering a suggestion as necessary.

Author: Makes sense! Updated in [75f55ca](https://github.com/MetaMask/core/pull/1976/commits/75f55ca445beb9f202134230f30396c7a442f146).

# Reviewing Release PRs

In order to publish a new release of the monorepo, a release PR must be created. But before a release PR is merged, it is absolutely critical — even if you are the author of the PR — to review the changes that are about to be published and to ensure that consumers know about them effectively.

This document shows you how to do that.

## 0. tl;dr — if you've done this sort of thing before

- [ ] Are there any packages with version bumps that should not be published?
- For all packages being published:
  - [ ] Does the new version string sufficiently communicate the impact of changes?
  - [ ] Are there any changelog entries that do not need to be listed?
  - [ ] Are there any changelog entries that are placed in the wrong category (including "Unreleased")?
  - [ ] Do all changelog entries explain the changes in a clear and straightforward manner and detail exact changes to the package's interface?

>   - [ ] Do all changes made through a pull request link to that pull request?

Reviewer: Suggested re-wording:

```suggestion
  - [ ] Do all change entries include a link to the pull requests where that change was made?
```

"Changes made through a pull request" seemed like a circuitous way of referring to all changes outside of the RC. It seems fine to me to just say "change entries", and let the author figure out whether this applies.

Better to not imply that each change entry corresponds to just a single PR as well. We can have entries that relate to multiple PRs.

>   - [ ] Are there any changes since the package's previous release that not reflected in the changelog?

Reviewer: Typo:
```suggestion
  - [ ] Are there any changes since the package's previous release that are not reflected in the changelog?
```

- Changes to interfaces
- Changes to behavior
- Changes to runtime dependencies
- Changes to consumer-facing documentation

Conversely, you generally should not need to worry about:

- Changes to internal tooling
- Changes to developer-only dependencies
- Changes to contributor-only documentation

With that in mind, there are four categories of changes that determine the level of impact.

### Breaking changes

If a consumer upgrades a package to a new version without making any modifications to their code, that version is "breaking" if it causes:

- An error at runtime
- An error at compile time (e.g., a TypeScript type error)
- An error at install time (e.g., the consumer's package manager reports a Node version incompatibility)
- A surprising difference in behavior

Given this, there are many ways a change could be regarded as breaking:

- Changing a function or method to throw an error
- Adding a required argument to a function or method

> - Adding a property to a TypeScript type

Reviewer: Thoughts? New optional properties are usually considered as non-breaking
```suggestion
- Adding a required property to a TypeScript type
```

Reviewer: I see that the same applies to the "Adding a parameter" point. Maybe we could consolidate both of these lines.

- Narrowing the type of a property in a TypeScript type
- Narrowing the type of an argument in a function
- Adding a parameter to a TypeScript type
- Removing a parameter from a TypeScript type
- Upgrading a runtime dependency to a version which causes any of the above
- Removing an export from the package
- Renaming the package

> - Bumping the Node version

Reviewer: To add a bit more detail:
```suggestion
- Bumping the minimum supported Node.js version
```

> - Bumping the NPM version

Reviewer: Hmm, what does this refer to?

Author: I believe this came from one of our forked repos where `engines.npm` was already specified in `package.json`, and we bumped it. That said, I've only seen this once and I suspect it will be very rare, so I can remove this.

Author: I have removed this item.

### Additions

Minimally, this category could cover any actions that extend the surface area of a package in some way:

- Adding an optional property to an existing type
- Adding an optional argument to an existing function or method
- Adding a new export to a package

> However, it is important to remember that consumers care about capabilities first and interfaces second. So, some additions could also be more simply regarded as changes. For instance:

Reviewer: I'm not entirely sure what "some additions could also be more simply regarded as changes" means here. The example given, adding a new dependency, doesn't suggest any "addition" to the package itself. Is that what you meant to communicate - that not every "addition" adds to the package?

Author: Yes. I'm trying to clarify that even if you are technically adding something new that wasn't there before, it doesn't mean that the change needs to be categorized as an addition. 

That said, it's confusing that this is here. I will figure out another place to put this.

Author: I updated the section on categories to fit the intersection between SemVer and Keep a Changelog and removed this confusing wording.

- Adding a new dependency

### Removals

In most cases, removing something from a package will cause a breaking change, and thus should be categorized as such. There are some rare instances, however, when this is not the case:

> - Removing a runtime dependency

Reviewer: Along similar lines, this is not something I'd view as a "removal" at all. Nothing is being removed from the package API when a dependency is removed.

Reviewer: In the context of deciding on what the SemVer version should be, I don't think we need to have this section here. Maybe we can remove this and let it be covered by the "Everything else" section

Author: > In the context of deciding on what the SemVer version should be, 

Hmm. The original goal of this section was not solely to decide what the SemVer version is, nor to decide how changes get categorized in the changelog. It was just to discuss different kinds of changes that you might see when preparing a release — what to pay attention to and what not to pay so much attention to — so that it was easier to figure out how to decide on a SemVer version and how to categorize changes in the changelog.

That said, I probably inadvertently mapped these categories map more to Keep a Changelog categories than SemVer categories. And in any case it seems confusing to have three separate ways of categorizing changes in this document.

I'll think about this some more.

Author: I updated the section on categories to fit the intersection between SemVer and Keep a Changelog. I removed this particular line but if there were any items that I wanted to call out as not belonging in a category, I marked them with an X.

### Fixes

There are a few different kinds of changes that could be filed under this category.

- Fixing a security vulnerability.
- Correcting a mistake introduced in a previous release. For example, a function was supposed to be included but was not.

> - Realigning confusing or illogical behavior. This is tricky, because the longer buggy behavior is allowed to persist, the more consumers may rely upon it. In this case, it would be more thoughtful to call out a change in this behavior as breaking, even if the new behavior makes more sense. (As an example, read about the [history of MetaMask's signing methods](https://github.com/MetaMask/metamask-docs/blob/72e1a066ea1a81a7964dbc9aa83381493a2cd3eb/wallet/concepts/signing-methods.md).) But for minor cases, a "fix" may be more appropriate.

Reviewer: The phrasing here seems to imply that "Breaking" and "Fixes" are mutually incompatible, but that's not really the case. Maybe this note about the difficulty in determining whether something is "breaking" doesn't belong in this section?

Reviewer: Oh sorry, they are incompatible in this context. I was thinking of the changelog entry categories. Nevermind this comment.

Now that you understand the changes being made and what sort of impact they are likely to have on consumers, the next step is to verify that the impact is being appropriately communicated via version strings.

[Semantic Versioning](https://semver.org/) (SemVer) dictates that versions should be bumped in the following manner:

- If there are any changes you would categorize as a breaking change, bump the **major** (first) part of the version.
- Otherwise, if there are any changes you would categorize as an addition, bump the **minor** (second) part of the version.
- Otherwise, bump the **patch** (third) part of the version.

If there are any packages being bumped whose new versions understate their impact, then notify the creator of the release PR so that they can update the versions appropriately.

## 4. Review existing changelog entries

Next, look over the changelogs in the packages-to-be-released to ensure that they are well formed and that they list all changes that will be released.

The [Keep a Changelog](https://keepachangelog.com/) specification defines a standard format for changelogs, and all MetaMask repositories, including the core monorepo, follow this format.

> You might notice that the categories listed in this spec match the same ones listed above in this document. This is not a coincidence!

Reviewer: Hmmm... the "Breaking" category certainly isn't one of the change entries.

Author: Yes, this statement is inaccurate. I can remove it.

Author: This is removed.
